lib/sidekiq/history/middleware.rb

```diff
@@ -10,56 +10,58 @@ def call(worker, msg, queue, &block)
       private
 
       def call_with_sidekiq_history(worker, msg, queue)
-        worker_status = new_status
+        worker_status = { last_job_status: 'passed'.freeze }
+
         start = Time.now.utc
 
         yield
       rescue StandardError => e
-        worker_status[:failed] = 1
-        worker_status[:passed] = 0
         worker_status[:last_job_status] = 'failed'.freeze
 
         raise e
       ensure
-        worker_status[:runtime] ||= [elapsed(start)]
+        worker_status[:time] = elapsed(start)
         worker_status[:last_runtime] = Time.now.utc
-        worker_status[:last_job_status] ||= 'passed'.freeze
         worker_status[:class] = msg['wrapped'.freeze] || worker.class.to_s
 
         save_entry_for_worker worker_status
       end
 
-      def new_status
-        {
-          failed: 0,
-          passed: 1
-        }
-      end
-
       def save_entry_for_worker(worker_status)
         status = worker_status.dup
-        worker = status.delete :class
+        worker_key = "#{Time.now.utc.to_date}:#{status.delete :class}"
 
         Sidekiq.redis do |redis|
-          history = "sidekiq:history:#{Time.now.utc.to_date}"
+          redis.watch(REDIS_HASH) do
+            min_time = hget_time(redis, "#{worker_key}:min_time", status)
+            max_time = hget_time(redis, "#{worker_key}:max_time", status)
+            average_time = hget_time(redis, "#{worker_key}:average_time", status)
 
-          redis.watch(history) do
-            value = redis.get history
+            statistics = [
+              "#{worker_key}:last_job_status", status[:last_job_status],
+              "#{worker_key}:average_time", (average_time + status[:time]) / 2,
+              "#{worker_key}:last_time", status[:last_runtime],
+              "#{worker_key}:min_time", [min_time, status[:time]].min,
+              "#{worker_key}:max_time", [max_time, status[:time]].max,
+            ]
 
             redis.multi do |multi|
-              if value && summary = Sidekiq.load_json(value)[worker]
-                summary = summary.symbolize_keys
-                [:failed, :passed, :runtime].each do |stat|
-                  status[stat] = summary[stat] + status[stat]
-                end
-              end
+              multi.hsetnx REDIS_HASH, "#{worker_key}:passed", 0
```

Comment:
Set the hash's `default` proc to return 0 instead of nil. 

> On Jul 18, 2015, at 03:32, Anton Davydov <notifications@github.com> wrote:
> 
> In lib/sidekiq/history/middleware.rb:
> 
> >  
> >              redis.multi do |multi|
> > -              if value && summary = Sidekiq.load_json(value)[worker]
> > -                summary = summary.symbolize_keys
> > -                [:failed, :passed, :runtime].each do |stat|
> > -                  status[stat] = summary[stat] + status[stat]
> > -                end
> > -              end
> > +              multi.hsetnx REDIS_HASH, "#{worker_key}:passed", 0
> This need for getting undefined keys. on client I am getting all keys and values for redis hash and if passed or failed is undefined - I am getting nil instead 0.
> 
> —
> Reply to this email directly or view it on GitHub.
>